### DISCUSSION TITLE
How to pass array of counter values into RegisterCallback()

### DISCUSSION THREAD

**Discussion by karthi8185** [open-telemetry/opentelemetry-go#4157] [General]
Hi, 
    I am using opentelemetry to export some customize data to otelcollector. Here my question is any possibility to pass the array arguments into RegisterCallback()

In the below code, instead of passing name1[0], name1[1], name1[2]... any possibility of send name1. Also, is there any api/method to resolve the type converstion. Since name1 data type is metric.Int64ObservableCounter but register to call back argument is metric.Observable.

var name1 []metric.Int64ObservableCounter
......
_, err = meter.RegisterCallback(                        
                                func(ctx context.Context, o metric.Observer) error {
                                  o.ObserveInt64(name1[0], int64(getMetricValue(fullName1)))
                                  o.ObserveInt64(name1[1], int64(getMetricValue(fullName2)))
                                  o.ObserveInt64(name1[2], int64(getMetricValue(fullName3)))
                                  o.ObserveInt64(name1[3], int64(getMetricValue(fullName4)))
                                return nil                                      
                                },                                              
                                name1[0],                                       
                                name1[1],                                       
                                name1[2],                                       
                                name1[3],                                       
                                )           

kindly guide me.

**Comment by pellared** [comment]
> is there any api/method to resolve the type converstion

No. But you can create a helper function yourself.

**Reply by karthi8185** [reply]
Can you pls explain a little bit more information about (steps), it's really helpful to understand. 
Since I am new to Golang. So I will try it from my end.

**Reply by dmathieu** [reply]
Something like this, which is pure go, not otel specific?

```golang
var name1 []metric.Int64ObservableCounter
_, err = meter.RegisterCallback(
    func(ctx context.Context, o metric.Observer) error {
        for i, v := range name1 {
            // getMetricValue is a method which knows the value to return based on `i` and `v` (the index and value of the observable counters). The actual logic/arguments of that method depends on what you need to do
            o.ObserveInt64(v, int64(getMetricValue(i, v)))
        }
        return nil
    }, name1...,
)
```

**Reply by karthi8185** [reply]
@dmathieu my current code is similar to that.. But i am getting below error

cannot use name1 (variable of type []"go.opentelemetry.io/otel/metric".Int64ObservableCounter) as []"go.opentelemetry.io/otel/metric".Observable value in argument to meter.RegisterCallback

As per my understanding type mismatch for the second argument (name1 here). 

 func (m *meter) RegisterCallback(f metric.Callback, insts ...metric.Observable) (metric.Registration, error)

**Reply by dmathieu** [reply]
You're trying to pass a slice of `Int64ObservableCounter` (see the `[]` at the beginning of the type name). You need to pass a single attribute.
The `...` I put in my example should turn the slice into arguments. See https://gobyexample.com/variadic-functions

**Reply by karthi8185** [reply]
Hi @dmathieu 
I did the same and got that error during compilation.

cannot use name1 (variable of type []"go.opentelemetry.io/otel/metric".Int64ObservableCounter) as []"go.opentelemetry.io/otel/metric".Observable value in argument to meter.RegisterCallback


 _, err = meter.RegisterCallback(                        
                                func(ctx context.Context, o metric.Observer) error {
                                  o.ObserveInt64(name1[0], int64(getMetricValue(fullName1)))
                                  o.ObserveInt64(name1[1], int64(getMetricValue(fullName2)))
                                  o.ObserveInt64(name1[2], int64(getMetricValue(fullName3)))
                                  o.ObserveInt64(name1[3], int64(getMetricValue(fullName4)))
                                return nil                                      
                                },                                              
                                name1...,                                       
                                )       

Please help on this.
